relay-server/src/actors/project.rs

```diff
@@ -595,6 +605,60 @@ impl Actor for Project {
     }
 }
 
+struct RateLimitEnvelope {
+    event_id: Option<EventId>,
+    remote_addr: Option<IpAddr>,
+    envelope_summary: EnvelopeSummary,
+    outcome_producer: Addr<OutcomeProducer>,
+    scoping: Scoping,
+}
+
+impl RateLimitEnvelope {
+    fn new(message: &CheckEnvelope, scoping: &Scoping) -> Self {
+        let envelope = &message.envelope;
+        Self {
+            event_id: envelope.event_id(),
+            remote_addr: envelope.meta().client_addr(),
+            envelope_summary: EnvelopeSummary::compute(&envelope),
+            outcome_producer: message.outcome_producer.clone(),
+            scoping: *scoping,
+        }
+    }
+
+    fn emit_rate_limit_outcomes(&self, applied_limits: &RateLimits) {
+        for applied_limit in applied_limits.iter() {
+            if applied_limit.categories.is_empty() {
+                // Empty categories value indicates that the rate limit applies to all data.
```

Comment:
Per https://github.com/getsentry/relay/blob/master/relay-quotas/src/rate_limit.rs#L140-L141. So far I have only blind faith in that comment to support that this is correct/necessary behavior, but maybe the integration tests will clarify things as I continue digging.

Reply:
I believe this approach is problematic since you cannot reconstruct the accurate drop reason from the `RateLimits` structure, unfortunately. This is an inherent flaw of `RateLimits` and we can consider changing that, too. 

To illustrate, consider the following example:
- There is a single quota `category:error limit:0`. It drops all error events.
- `EnvelopeLimiter` will also drop all attachments as a result of that.
- When you check `RateLimits` here, you won't find the `attachment` category, even though you just dropped them. 

From the top of my head, I have two ideas to solve that:
1. Move emitting outcomes into `EnvelopeLimiter`, because that's where you can keep track of the dropped quantities.
2. From `EnvelopeLimiter::enforce`, return an instance of `EnvelopeSummary` that contains the dropped quantities. Then, use that summary (+ scoping) to emit outcomes. This approach is preferable as it decouples concerns.